jlm/hls/opt/InvariantLambdaMemoryStateRemoval.cpp

```diff
@@ -0,0 +1,157 @@
+  if (!rvsdg::is<const llvm::LambdaExitMemoryStateMergeOperation>(exitNode->GetOperation()))
+  {
+    return;
+  }
+
+  // Check if we have any invariant edge(s) between the two nodes
+  std::vector<size_t> indices;
+  std::vector<rvsdg::output *> nonInvariantOutputs;
+  rvsdg::Node * entryNode = nullptr;
+  for (size_t i = 0; i < exitNode->ninputs(); i++)
+  {
+    // Check if the output has only one user and if it is a LambdaEntryMemoryStateMerge
+    auto node = rvsdg::TryGetOwnerNode<rvsdg::Node>(*exitNode->input(i)->origin());
+    if (jlm::rvsdg::is<const llvm::LambdaEntryMemoryStateSplitOperation>(node->GetOperation()))
+    {
+      // Found an invariant memory state edge, so going to replace the entryNode
+      entryNode = node;
+    }
+    else
+    {
+      // Keep track of edges that is to be kept since they are not invariant
+      nonInvariantOutputs.push_back(exitNode->input(i)->origin());
+      // Also keep track of the index to be used for diverting edges
+      indices.push_back(i);
+    }
+  }
+
+  // If the entryNode is not set, then we haven't found any invariant edges
+  if (entryNode == nullptr)
+  {
+    return;
+  }
+
+  // Replace LambdaEntryMemoryStateSplit and LambdaExitMemoryStateMerge nodes
+  if (nonInvariantOutputs.size() == 0)
+  {
+    // The memory state edge(s) are invariant, so we could in principle remove all of them from the
+    // lambda But the LLVM dialect expects to always have a memory state, so we connect the argument
+    // directly to the result
+    memoryStateResult->divert_to(entryNode->input(0)->origin());
+  }
+  else if (nonInvariantOutputs.size() == 1)
+  {
+    // Single edge that is not invariant, so we can elmintate the two MemoryState nodes
+    memoryStateResult->divert_to(nonInvariantOutputs[0]);
+    entryNode->output(indices[0])->divert_users(entryNode->input(0)->origin());
+  }
+  else
+  {
+    // Replace the entry and exit node with new ones without the invariant edge(s)
+    auto newEntryNodeOutputs = llvm::LambdaEntryMemoryStateSplitOperation::Create(
+        *entryNode->input(0)->origin(),
+        indices.size());
+    memoryStateResult->divert_to(&llvm::LambdaExitMemoryStateMergeOperation::Create(
+        *exitNode->region(),
+        nonInvariantOutputs));
+    int i = 0;
+    for (auto index : indices)
+    {
+      entryNode->output(index)->divert_users(newEntryNodeOutputs[i]);
+      i++;
+    }
+  }
+  JLM_ASSERT(exitNode->IsDead());
+  rvsdg::remove(exitNode);
+  JLM_ASSERT(entryNode->IsDead());
+  rvsdg::remove(entryNode);
+}
+
+void
+InvariantLambdaMemoryStateRemoval::RemoveInvariantLambdaMemoryStateEdges(
+    rvsdg::RvsdgModule & rvsdgModule)
+{
+  for (auto & node : rvsdg::Graph::ExtractTailNodes(rvsdgModule.Rvsdg()))
+  {
+    if (auto lambda = dynamic_cast<const rvsdg::LambdaNode *>(node))
+    {
+      if (lambda->output()->nusers() == 1
+          && dynamic_cast<const jlm::rvsdg::GraphExport *>(*lambda->output()->begin()))
+      {
```

Comment:
You would like to add another check here that the `LambdaNode` carries a `llvm::LambdaOperation` (and not a `rvsdg::LambdaOperation`). Otherwise, you have no guarantees that a memory state argument exists in the first place.

Reply:
Added